util/rateLimit.go

```diff
@@ -0,0 +1,470 @@
+type redCongestionManager struct {
+	runLock                *deadlock.Mutex
+	running                bool
+	window                 time.Duration
+	consumed               chan event
+	served                 chan event
+	shouldDropQueries      chan shouldDropQuery
+	targetRate             float64
+	targetRateRefreshTicks int
+	// exp is applied as an exponential factor in shouldDrop. 1 would be linearly proportional, higher values punish noisy neighbors more
+	exp float64
+	// consumed is the only value tracked by-queue. The others are calculated in-total
+	// TODO: If we desire later, we can add mappings onto release/done for more insight
+	consumedByClient map[ErlClient]*[]time.Time
+	serves           []time.Time
+}
+
+// NewREDCongestionManager creates a Congestion Manager which will watches capacityGuard activity,
+// and regularly calculates a Target Service Rate, and can give "Should Drop" suggestions
+func NewREDCongestionManager(d time.Duration, r int) *redCongestionManager {
+	ret := redCongestionManager{
+		runLock:                &deadlock.Mutex{},
+		window:                 d,
+		consumed:               make(chan event, 100000),
+		served:                 make(chan event, 100000),
+		shouldDropQueries:      make(chan shouldDropQuery, 100000),
+		targetRateRefreshTicks: r,
+		consumedByClient:       map[ErlClient]*[]time.Time{},
+		exp:                    4,
+	}
+	return &ret
+}
+
+// Consumed implements CongestionManager by putting an event on the consumed channel,
+// to be processed by the Start() loop
+func (cm redCongestionManager) Consumed(c ErlClient, t time.Time) {
+	select {
+	case cm.consumed <- event{
+		c: c,
+		t: t,
+	}:
+	default:
+	}
+}
+
+// Served implements CongestionManager by putting an event on the done channel,
+// to be processed by the Start() loop
+func (cm redCongestionManager) Served(t time.Time) {
+	select {
+	case cm.served <- event{
+		t: t,
+	}:
+	default:
+	}
+}
+
+// ShouldDrop implements CongestionManager by putting a query shouldDropQueries channel,
+// and blocks on the response to return synchronously to the caller
+// if an error should prevent the query from running, the result is defaulted to false
+func (cm redCongestionManager) ShouldDrop(c ErlClient) bool {
+	ret := make(chan bool)
+	select {
+	case cm.shouldDropQueries <- shouldDropQuery{
+		c:   c,
+		ret: ret,
+	}:
+		return <-ret
+	default:
+		return false
+	}
+}
+
+// Start will kick off a goroutine to consume activity from the different activity channels,
+// as well as service queries about if a given capacityQueue should drop
+func (cm *redCongestionManager) Start(ctx context.Context, wg *sync.WaitGroup) {
+	// check if the maintainer is already running to ensure there is only one routine
+	cm.runLock.Lock()
+	defer cm.runLock.Unlock()
+	if cm.running {
+		return
+	}
+	cm.running = true
+	go cm.run(ctx, wg)
+}
+
+func (cm *redCongestionManager) run(ctx context.Context, wg *sync.WaitGroup) {
```

Comment:
looks like wg *sync.WaitGroup is not used by production code, get rid of it?

Reply:
it's used to make sure the structure is done running before inspecting via unit tests. without this, we get race conditions when asserting on the congestion manager

Reply:
talked offline -- will attach the wg to the struct directly so it doesn't have to be passed around like this